Fix order notes not persisting

alexflorisca opened the pull request:
### Changes proposed in this Pull Request:

This fixes a bug where the order notes don't persist, following some changes to how checkout fields persist. The problem was a mismatch between the keys used on the client and server to reder to the order notes. The client was sending `order_notes` and the server was expecting `customer_note`. We are now using `customer_note` for both, and only updating the customer note if it is being appended to the request.

Also updated a test to catch this which was missing.

Update by @<!-- linear:userMention:a11300a5-c516-4ff0-a705-73279ff7c669 -->thomas.roberts  - Also added code to remove the order notes when unchecking the Order Notes checkbox. This brings it more in line with the rest of the AdditionalCheckout Fields API conditional fields.

Closes woocommerce/woocommerce#56555.

(For Bug Fixes) Bug introduced in PR [#55989](<https://github.com/woocommerce/woocommerce/issues/55989>).

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](<https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions>), include your detailed testing instructions:

1. Install and activate the [Additional Fields Tester](<https://github.com/woocommerce/additional-checkout-fields-tester>) plugin.
2. Add an item to your cart and go to the Checkout block
3. Tick the order notes box and add some text.
4. Wait 1.5s or more (or wit for a PUT request to /checkout then refresh the page.
5. Check that what you entered in the order notes is still there
6. Update one of the additional fields, wait as in step 4 and refresh
7. Check that the order notes are still there
8. Update all the additional fields, order notes and payment method, refresh, and check they are persisted.

### Testing that has already taken place:

### Changelog entry

- [X] Automatically create a changelog entry from the details below.
- [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details><summary>Changelog Entry Details</summary>

#### Significance

- [X] Patch
- [ ] Minor
- [ ] Major

#### Type

- [ ] Fix - Fixes an existing bug
- [X] Add - Adds functionality
- [ ] Update - Update existing functionality
- [ ] Dev - Development related task
- [ ] Tweak - A minor adjustment to the codebase
- [ ] Performance - Address performance issues
- [ ] Enhancement - Improvement to existing functionality

#### Message

Persist order notes in checkout block

</details>

<details><summary>Changelog Entry Comment</summary>

#### Comment

</details>

Commit:
Fix order notes not persisting

Commit:
Add test to check for order notes

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Commit:
Merge branch 'trunk' into fix/order-notes-not-persisting

opr commented on plugins/woocommerce/src/StoreApi/Utilities/CheckoutTrait.php:
```diff
@@ -146,7 +146,9 @@ private function get_request_payment_data( \WP_REST_Request $request ) {
 	private function update_order_from_request( \WP_REST_Request $request ) {
-		$this->order->set_customer_note( wc_sanitize_textarea( $request['customer_note'] ) ?? '' );
+		if ( isset( $request['customer_note'] ) ) {
+			$this->order->set_customer_note( wc_sanitize_textarea( $request['customer_note'] ) ?? '' );
```
Minor improvement/nitpick, but null coalesce is not needed here since we're checking if `isset` before coming here. Null coalesce will never happen. 

Removing this could potentially improve code clarity, and prevent a future engineer looking into why the null coalesce is needed.

alexflorisca replied:
Great find, done

opr reviewed (dismissed):
Hi @alexflorisca thanks for working on this, looking good, tested on JN and locally and both working OK. I asked a question (See below in the last testing step I carried out, about refreshing after closing the order notes box) that should be answered and clarified, possibly involving product/design before merging ([taking a learning from my escape analysis here](https://github.com/woocommerce/woocommerce/issues/56529#issuecomment-2737330821)).

## Testing Guidelines

Apart from reviewing the code changes, please make sure to **review the testing instructions** ([Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)) and verify that relevant tests (E2E, Unit, Integration, etc.) have been added or updated as needed.

**Reminder:** PR reviewers are required to document testing performed. This includes:
- 🖼️ **Screenshots or screen recordings**.
  - https://www.loom.com/share/2d4be3a8c81f4781ad89ee2bef683c81
- 📝 **List of functionality tested / steps followed**.
  - Followed the steps in the PR body.
  - Did some edge-case testing.
    - Modify custom field, modify order notes, wait for PUT, then reload. Expected: All are persisted after reload.
    - Modify order notes without modifying any custom field, wait for PUT, then reload. Expected: Notes are persisted after reload.
    - Modify custom field, modify order notes, modify a different custom field, wait for PUT, then reload. Expected: All changes are persisted after reload.
    - Modify order notes, wait for PUT. Uncheck `Add a note to your order`. Wait for PUT. Re-check `Add a note to your order`. Expect notes to remain. 
    - Modify order notes, wait for PUT. Uncheck `Add a note to your order`. Wait for PUT. Modify a custom field. Wait for PUT. Re-check `Add a note to your order`. Expect notes to remain. 
    - Modify order notes, wait for PUT. Uncheck `Add a note to your order`. Wait for PUT. Reload. Re-check `Add a note to your order`. Expect notes to be blank - at least that is my expectation as a developer. **Would a shopper have the same expectation? If you uncheck order notes, reload, and re-check it, would you realistically expect them to be there?** 👀 @alexflorisca 

I also tested this _without_ the additional fields tester plugin, and as a logged-out user in Firefox.

- 🌐 **Site details** (environment attributes such as hosting type, plugins, theme, store size, store age, and relevant settings).
  - Tested with Local WP
<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 9.9.0-dev
Legacy REST API Package Version: The Legacy REST API plugin is not installed on this site.
Action Scheduler Version: ✔ 3.9.2
Log Directory Writable: ✔
WP Version: 6.8-beta3-60060
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 8.2.8
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 8.7.1
(SecureTransport) LibreSSL/3.3.6

SUHOSIN Installed: –
MySQL Version: 8.0.16
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

af_c_fields: 1
attachment: 37
aw_workflow: 1
custom_css: 1
def_reg_fields: 10
jetpack_migration: 2
jp_img_sitemap: 1
jp_sitemap: 1
jp_sitemap_master: 1
page: 19
post: 184
product: 33
product_variation: 48
revision: 709
shop_coupon: 13
shop_order: 704
shop_order_placehold: 16
shop_order_refund: 1
shop_subscription: 1
wc_booking: 1
wc_conditional_fee: 1
wc_cs_form_field: 12
woo-docs-manifest: 1
wp_font_face: 36
wp_font_family: 12
wp_global_styles: 7
wp_navigation: 1
wp_template: 4
wp_template_part: 8
yaymail_template: 15

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (19) ###

Action Scheduler: by Automattic – 3.9.2
Additional Checkout Fields Tester: by WooCommerce – 1.0
Code Snippets: by Code Snippets Pro – 3.6.5.1
Gutenberg: by Gutenberg Team – 20.3.0-rc.1
Jetpack: by Automattic – 13.9.1
JSM Show Post Metadata: by JS Morisset – 4.6.0
Query Monitor: by John Blackbourn – 3.17.2
Restore Paypal Standard For WooCommerce: by Jose Mortellaro – 1.0.3
Sandbox: by The WordPress Contributors – 0.1.0
WooCommerce Beta Tester: by WooCommerce – 3.0.0
WooCommerce Checkout Field Editor: by WooCommerce – 1.7.14 (update to version 1.7.22 is available)
WooCommerce Composite Products: by Woo – 10.2.0 (update to version 10.3.4 is available)
WooCommerce Docs: by Automattic – 1.1
WooCommerce Stripe Gateway: by Stripe – 9.3.1
WooCommerce Product Add-Ons: by Woo – 7.1.2 (update to version 7.7.0 is available)
WooCommerce Shipping Per Product v2: by WooCommerce – 2.6.1 (update to version 2.6.5 is available)
WooCommerce Square: by WooCommerce – 4.8.7
WooCommerce: by Automattic – 9.9.0-dev
WordPress Beta Tester: by WordPress Upgrade/Install Team – 3.6.2

### Inactive Plugins (53) ###

Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.3.7
AutomateWoo: by WooCommerce – 6.1.1 (update to version 6.1.8 is available)
Core Rollback: by Andy Fragen – 1.3.5
Create Block Theme: by WordPress.org – 1.13.8
Currency Switcher for WooCommerce: by WP Wham – 2.15.2
Custom Fields for WooCommerce: by Addify – 1.3.0
Debug Log Manager: by Bowo – 2.3.3
Dogfood: by Thomas
Nadir
Paulo – 1.0

Example add notice: by Thomas – 1.0
Example Block: by The WordPress Contributors – 0.1.0
Example Incompatible Plugin: by The WordPress Contributors – 0.1.0
Example Static (ES5): by The WordPress Contributors – 0.1.0
Express Checkout via PayPal for WooCommerce: by wpgateways – 6.0.0
Failed Gateway: by It could be anyone. – 1.0
Fake Carousel: by The WordPress Contributors – 0.1.0
jQuery Updater: by Ramoonus – 3.7.1.3
Loco Translate: by Tim Whitlock – 2.6.9
Multiple Packages for WooCommerce: by Erica Dion – 1.1.1
Page Optimize: by Automattic – 0.5.5
Polylang: by WP SYNTEX – 3.6.6
Product Table by WBW: by woobewoo – 2.1.0 (update to version 2.1.4 is available)
Redis Object Cache: by Till Krüss – 2.5.1
Rental Products: by 99w – 4.4.1 (update to version 6.0.0 is available)
Shipping Workshop: by Thomas Roberts and Niels Lange – 1.0
Shipping Workshop: by Thomas Roberts and Niels Lange – 1.0
Simple Custom CSS and JS: by SilkyPress.com – 3.46
Store API Quantity Backward Compatibility: by Thomas Roberts – 0.1.0
Test Extension Cart Update: by The WordPress Contributors – 0.1.0
Transients Manager: by WPBeginner – 2.0.5
VIP NotOptions Mitigation: by Automattic –
W3 Total Cache: by BoldGrid – 2.7.5
WooCommerce Conditional Shipping and Payments: by WooCommerce – 1.15.3 (update to version 2.2.0 is available)
WooCommerce Deposits: by WooCommerce – 2.2.8 (update to version 2.3.7 is available)
WooCommerce Dummy Payments Gateway: by SomewhereWarm – 1.0.8
WooCommerce EU VAT Number: by WooCommerce – 2.9.4 (update to version 2.9.14 is available)
WooCommerce Extra Fees Plugin: by theDotstore – 3.9.5
WooCommerce Google Analytics Integration: by WooCommerce – 1.8.11
WooCommerce Legacy REST API: by WooCommerce – 1.0.4
WooCommerce Min/Max Quantities: by Woo – 5.1.0 (update to version 5.2.3 is available)
WooCommerce Mix and Match Products: by Kathy Darling – 2.7.1 (update to version 2.7.9 is available)
WooCommerce One Page Checkout: by Automattic – 2.8.2 (update to version 2.9.5 is available)
WooCommerce PayPal Payments: by WooCommerce – 2.8.1 (update to version 3.0.0 is available)
WooCommerce Points and Rewards: by WooCommerce – 1.8.2 (update to version 1.8.16 is available)
WooCommerce Product Bundles: by WooCommerce – 6.17.3 (update to version 8.3.4 is available)
WooCommerce Readonly Address Fields: by WooCommerce – 0.1.0
WooCommerce Square: by WooCommerce – 4.8.7
WooCommerce Stripe Gateway: by Stripe – 9.3.1
WooCommerce Subscriptions Core: by Automattic – 7.7.1
Woo Subscriptions: by WooCommerce – 6.1.0 (update to version 7.2.1 is available)
WP-Optimize - Clean, Compress, Cache: by David Anderson
Ruhani Rabin
Team Updraft – 3.2.19

WP Hooks Finder: by Muhammad Rehman – 1.3.2
Yoast SEO: by Team Yoast – 24.5
ZExample Dynamic Block: by The WordPress Contributors – 0.1.0

### Dropin Plugins () ###

db.php: Query Monitor Database Class (Drop-in)

### Settings ###

Legacy API Enabled: –
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: booking (booking)
bundle (bundle)
composite (composite)
external (external)
grouped (grouped)
mix and match (mix-and-match)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔
Enforce Approved Product Download Directories: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 552 KB

### WC Pages ###

Shop base: #6 - /shop/
Cart: #1510 - /cart/ -  Contains the woocommerce/cart block
Checkout: #1076 - /checkout/ -  Contains the woocommerce/checkout block
My account: #9 - /my-account/ -  Contains the [woocommerce_my_account] shortcode
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Five
Version: 1.1
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
Theme type: Block theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: /Users/thomas/code/tomato/app/public/wp-content/plugins/woocommerce-monorepo/plugins/woocommerce/templates/block-notices/error.php
/Users/thomas/code/tomato/app/public/wp-content/plugins/woocommerce-monorepo/plugins/woocommerce/templates/block-notices/notice.php
/Users/thomas/code/tomato/app/public/wp-content/plugins/woocommerce-monorepo/plugins/woocommerce/templates/block-notices/success.php


### WooCommerce Stripe Payment Gateway ###

Version: 9.3.1
Account ID: acct_1R4j8aLmlx3dgeMq
Account Email: Unknown
Test Mode Enabled: Yes
OAuth Connected: ✔
					Yes
Legacy Checkout Experience: ✔
					No
Enabled Payment Methods: card,link
Express Checkout: ✔
					Enabled (product,cart,checkout)
Auth and Capture: ✔
					Yes
Logging: No

### Composite Products ###

Database Version: 10.2.0
Loopback Test: ✔
Template Overrides: –

### Square ###

Environment: Production
Tokenization Enabled: ✔
Debug Mode: Display at Checkout & Log

### Cash App Pay (Square) ###

Environment: Production
Debug Mode: Display at Checkout & Log

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
experimental-blocks
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
minified-js
mobile-app-banner
onboarding
onboarding-tasks
pattern-toolkit-full-composability
payment-gateway-suggestions
product-custom-fields
printful
remote-inbox-notifications
remote-free-extensions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
async-product-editor-category-field
launch-your-store
add-to-cart-with-options-stepper-layout
blockified-add-to-cart
disable-core-profiler-fallback

Disabled Features: product-data-views
coming-soon-newsletter-template
product-pre-publish-modal
settings
product-editor-template-system
reactify-classic-payments-settings
use-wp-horizon
beta-tester-slotfill-examples

Daily Cron: ✔ Next scheduled: 2025-03-22 13:32:22 +00:00
Options: ✔
Notes: 151
Onboarding: completed

### Action Scheduler ###

Complete: 1,201
Oldest: 2025-02-18 14:56:59 +0000
Newest: 2025-03-21 14:46:04 +0000

Failed: 187
Oldest: 2023-07-07 13:36:09 +0000
Newest: 2025-03-20 15:45:08 +0000

In-progress: 1
Oldest: –
Newest: –

Pending: 1
Oldest: 2025-03-21 15:31:37 +0000
Newest: 2025-03-21 15:31:37 +0000


### Status report information ###

Generated at: 2025-03-21 14:47:49 +00:00
```
</details>

  - Tested on JN with live branches

<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 9.9.0
Legacy REST API Package Version: The Legacy REST API plugin is not installed on this site.
Action Scheduler Version: ✔ 3.9.2
Log Directory Writable: ❌ To allow logging
make /srv/htdocs/wp-content/uploads/wc-logs/ writable.

WP Version: 6.7.2
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: ✔

### Server Environment ###

Server Info: nginx
PHP Version: 8.2.28
PHP Post Max Size: 2 GB
PHP Time Limit: 850
PHP Max Input Vars: 6144
cURL Version: 8.12.1
OpenSSL/3.0.15

SUHOSIN Installed: –
MySQL Version: 10.11.11-MariaDB-log
Max Upload Size: 2 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

attachment: 11
page: 7
post: 3
product: 9
revision: 1
shop_order_placehold: 1
wp_navigation: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (3) ###

Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.3.7
WooCommerce: by Automattic – 9.9.0-dev-13976761861-g04fdab0
WooCommerce Beta Tester: by WooCommerce – 3.0.0

### Inactive Plugins (0) ###


### Dropin Plugins () ###

advanced-cache.php: advanced-cache.php
object-cache.php: Memcached

### Must Use Plugins (1) ###

load.php: by  –

### Settings ###

Legacy API Enabled: –
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 0 B

### WC Pages ###

Shop base: #5 - /shop/
Cart: #6 - /cart/ -  Contains the woocommerce/cart block
Checkout: #7 - /checkout/ -  Contains the woocommerce/checkout block
My account: #8 - /my-account/ -  Contains the [woocommerce_my_account] shortcode
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Four
Version: 1.3
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
Theme type: Block theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: /srv/htdocs/wp-content/plugins/wc_beta_tester_live_branch_9.9.0-dev-13976761861-g04fdab0/templates/block-notices/error.php
/srv/htdocs/wp-content/plugins/wc_beta_tester_live_branch_9.9.0-dev-13976761861-g04fdab0/templates/block-notices/notice.php
/srv/htdocs/wp-content/plugins/wc_beta_tester_live_branch_9.9.0-dev-13976761861-g04fdab0/templates/block-notices/success.php


### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
onboarding
onboarding-tasks
pattern-toolkit-full-composability
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
printful
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store
add-to-cart-with-options-stepper-layout
disable-core-profiler-fallback

Disabled Features: product-data-views
experimental-blocks
coming-soon-newsletter-template
minified-js
product-pre-publish-modal
settings
async-product-editor-category-field
product-editor-template-system
reactify-classic-payments-settings
use-wp-horizon
blockified-add-to-cart
beta-tester-slotfill-examples

Daily Cron: ✔ Next scheduled: 2025-03-22 15:06:51 +00:00
Options: ✔
Notes: 58
Onboarding: skipped

### Action Scheduler ###

Complete: 31
Oldest: 2025-03-21 15:07:52 +0000
Newest: 2025-03-21 15:09:00 +0000

Pending: 1
Oldest: 2025-03-22 15:07:52 +0000
Newest: 2025-03-22 15:07:52 +0000


### Status report information ###

Generated at: 2025-03-21 15:09:52 +00:00
```
</details>

- 🔍 **Any analysis performed**, such as assessing potential impacts on environment attributes and other plugins, conducting performance profiling, or using LLM/AI-based analysis.

I used this tool (p7H4VZ-5ff-p2) by @zmaglica (👏🏼) to help me generate a prompt for Claude.

Claude highlighted that the null coalesce is possibly unneeded and I looked into it and agreed. I explored with Claude the performance implications of keeping it (negligible, since it generates only a tiny bit more bytecode) but agreed that removing it may be better to prevent future engineers trying to understand its purpose.

Claude also mentioned back compatibility as an issue, e.g. people using `order_notes` but this is not a concern since the code is unreleased.

⚠️ Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public issue.

Confirmed 👍🏼

Commit:
Remove nullish coalescing operator

Commit:
Merge branch 'fix/order-notes-not-persisting' of github.com:woocommerce/woocommerce into fix/order-notes-not-persisting

alexflorisca commented:
> I'm curious why it was originally implemented as a checkbox instead of just a text field that if populated, is persisted and added to the order

Honestly not sure why, was before my time I think. @senadir or @mikejolley might know? It would actually improve the UI (1 less click) and simplify the logic here if we removed the checkbox.

senadir commented:
The design predates me but my removing the checkbox we are “simplifying” for a small subset of customers who leave a note, the majority of them, the checkbox visual length increases and it might be confused for a mandatory field. I can pull numbers later for the split of people who use notes

alexflorisca commented:
> Otherwise, per my previous comment @alexflorisca the state and input string I feel are best persisted independently of each other to handle this edge case. Do you agree?

We only care about persisting the text from the order notes text area. The backend order only has one field for order notes so I wouldn't want to complicate that server side. The checkbox is just to optimize front end like Nadir said, reducing the page length. So we don't need to persist that. 

I would say this is a very unlikely scenario. Someone would have to clear the notes AND reload the page for them to disappear. I don't want to complicate it. Considering it again with what Nadir said, I would opt to leave it as is.

alexflorisca commented:
Any more thoughts on this before I merge? @pierorocca @yjailin

yjailin commented:
My first reaction was to remove the checkbox and expose the text field directly, but i agree the checkbox is a good way to reduce scrolling. So i am okay keeping the checkbox as it is.

opr committed:
Remove persisted order notes when checkbox is unchecked

This will improve consistency with the rest of the hidden fields added through Additional Checkout Fields API

opr committed:
Update JS test to ensure order notes are cleared on uncheck

opr reviewed (dismissed):
@alexflorisca thanks again for your work here.

We are now going to deliberately clear the order notes field when the checkbox is unchecked to align the solution with the Additional Checkout Fields API where the same happens.

The notes should persist as expected when reloading the page, but should not persist at all when unchecking/checking the checkbox, even with no reload.

Here's this working:

https://www.loom.com/share/0461fc469a0d43dea8e0857d5d7f0eef

cc @senadir
